Compile failure after 0.9.6

I'm seeing a compile failure after upgrading to 0.9.6 on Linux. I'm a bit unsure, if this might have happened together with another upgrade. But this fails to compile using both clang and gcc. The offending code is:
 ```python
#pythran export train_equaliser(complex128[][], int, int, int, float64, complex128[][][], int[], bool, complex128[][], str)
#pythran export train_equaliser(complex64[][], int, int, int, float32, complex64[][][], int[], bool, complex64[][], str)
def train_equaliser(E, TrSyms, Niter, os, mu, wx, modes, adaptive, symbols,  method):
    if method == "cma":
        errorfct = cma_error

    else:
        raise ValueError("Unknown method %s"%method)
    nmodes = E.shape[0]
    ntaps = wx.shape[-1]
    assert symbols.shape[0] == nmodes, "symbols must be at least size of modes"
    assert wx.shape[0] == nmodes, "wx needs to have at least as many dimensions as the maximum mode"
    assert E.shape[1] > TrSyms*os+ntaps, "Field must be longer than the number of training symbols"
    assert modes.max() < nmodes, "Maximum mode number must not be higher than number of modes"
    err = np.zeros((nmodes, TrSyms*Niter), dtype=E.dtype)
    #omp parallel for
    for mode in modes:
        for it in range(Niter):
            for i in range(TrSyms):
                X = E[:, i * os:i * os + ntaps]
                Xest = apply_filter(X,  wx[mode])
                err[mode, it*Niter+i] = errorfct(Xest, symbols[mode], i)
                wx[mode] += mu * np.conj(err[mode, it*Niter+i]) * X
                if adaptive and i > 0:
                    mu = adapt_step(mu, err[mode, it*Niter+i], err[mode, it*Niter+i-1])
    return err, wx, mu

def cma_error(Xest, s1, i):
    d = s1[0].real - abs(Xest)**2
    return d*Xest
```
The compilation error I get is:
```
/home/jschrod/.local/lib/python3.8/site-packages/pythran/pythonic/utils/broadcast_copy.hpp:369:31: error: no matching function for call to ‘{anonymous}::pythonic::types::numpy_iexpr<const {anonymous}::pythonic::types::ndarray<std::complex<float>, {anonymous}::pythonic::types::pshape<long int, long int, long int> >&>::update<{anonymous}::pythonic::operator_::functor::iadd>(std::complex<double>, std::__tuple_element_t<0, std::tuple<long int, long int> >, std::__tuple_element_t<1, std::tuple<long int, long int> >)’
  369 |       self.template update<Op>(other.load((long)std::get<Is>(other_indices)...),
      |       ~~~~~~~~~~~~~~~~~~~~~~~~^~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
  370 |                                (long)std::get<Is>(self_indices)...);
      |                                ~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
In file included from /home/jschrod/.local/lib/python3.8/site-packages/pythran/pythonic/include/types/ndarray.hpp:38,
                 from /home/jschrod/.local/lib/python3.8/site-packages/pythran/pythonic/types/ndarray.hpp:4,
                 from /home/jschrod/.local/lib/python3.8/site-packages/pythran/pythonic/types/tuple.hpp:13,
                 from /home/jschrod/.local/lib/python3.8/site-packages/pythran/pythonic/builtins/bool_.hpp:7,
                 from /home/jschrod/.local/lib/python3.8/site-packages/pythran/pythonic/types/NoneType.hpp:8,
                 from /home/jschrod/.local/lib/python3.8/site-packages/pythran/pythonic/types/slice.hpp:5,
                 from /home/jschrod/.local/lib/python3.8/site-packages/pythran/pythonic/core.hpp:29,
                 from /tmp/tmpdeqs6i45.cpp:1:
```

If I change the line with `wx[mode] += mu*np.conj(err[mode, it*Niter+i])*X` to `wx[mode] = wx[mode] + mu*np.conj(err[mode, it*Niter+i])*X` it compiles again.

--- a/pythran/pythonic/include/types/ndarray.hpp
+++ b/pythran/pythonic/include/types/ndarray.hpp
@@ -338,11 +338,13 @@ namespace types
     template <class Expr>
     ndarray &operator^=(Expr const &expr);
 
-    template <class... Indices>
-    void store(dtype elt, Indices... indices)
+    template <class E, class... Indices>
+    void store(E elt, Indices... indices)
     {
+      static_assert(is_dtype<E>::value, "valid store");
       *(buffer + noffset<std::tuple_size<pS>::value>{}(
-                     *this, array<long, value>{{indices...}})) = elt;
+                     *this, array<long, value>{{indices...}})) =
+          static_cast<E>(elt);
     }
     template <class... Indices>
     dtype load(Indices... indices) const
@@ -351,12 +353,13 @@ namespace types
                             *this, array<long, value>{{indices...}}));
     }
 
-    template <class Op, class... Indices>
-    void update(dtype elt, Indices... indices) const
+    template <class Op, class E, class... Indices>
+    void update(E elt, Indices... indices) const
     {
+      static_assert(is_dtype<E>::value, "valid store");
       Op{}(*(buffer + noffset<std::tuple_size<pS>::value>{}(
                           *this, array<long, value>{{indices...}})),
-           elt);
+           static_cast<E>(elt));
     }
 
     /* element indexing

--- a/pythran/pythonic/include/types/numpy_gexpr.hpp
+++ b/pythran/pythonic/include/types/numpy_gexpr.hpp
@@ -767,11 +767,12 @@ namespace types
       return numpy_iexpr_helper<numpy_gexpr, value>::get(*this, i);
     }
 
-    template <class... Indices>
-    void store(dtype elt, Indices... indices)
+    template <class E, class... Indices>
+    void store(E elt, Indices... indices)
     {
+      static_assert(is_dtype<E>::value, "valid store");
       *(buffer + noffset<value>{}(*this, array<long, value>{{indices...}})) =
-          elt;
+          static_cast<E>(elt);
     }
     template <class... Indices>
     dtype load(Indices... indices) const
@@ -779,12 +780,13 @@ namespace types
       return *(buffer +
                noffset<value>{}(*this, array<long, value>{{indices...}}));
     }
-    template <class Op, class... Indices>
-    void update(dtype elt, Indices... indices) const
+    template <class Op, class E, class... Indices>
+    void update(E elt, Indices... indices) const
     {
+      static_assert(is_dtype<E>::value, "valid store");
       Op{}(
           *(buffer + noffset<value>{}(*this, array<long, value>{{indices...}})),
-          elt);
+          static_cast<E>(elt));
     }
 
 #ifdef USE_XSIMD

--- a/pythran/pythonic/include/types/numpy_iexpr.hpp
+++ b/pythran/pythonic/include/types/numpy_iexpr.hpp
@@ -199,11 +199,12 @@ namespace types
         numpy_vexpr<numpy_iexpr, ndarray<long, pshape<long>>>>::type
     fast(F const &filter) const;
 
-    template <class... Indices>
-    void store(dtype elt, Indices... indices)
+    template <class E, class... Indices>
+    void store(E elt, Indices... indices)
     {
+      static_assert(is_dtype<E>::value, "valid store");
       *(buffer + noffset<value>{}(*this, array<long, value>{{indices...}})) =
-          elt;
+          static_cast<E>(elt);
     }
     template <class... Indices>
     dtype load(Indices... indices) const
@@ -211,12 +212,13 @@ namespace types
       return *(buffer +
                noffset<value>{}(*this, array<long, value>{{indices...}}));
     }
-    template <class Op, class... Indices>
-    void update(dtype elt, Indices... indices) const
+    template <class Op, class E, class... Indices>
+    void update(E elt, Indices... indices) const
     {
+      static_assert(is_dtype<E>::value, "valid store");
       Op{}(
           *(buffer + noffset<value>{}(*this, array<long, value>{{indices...}})),
-          elt);
+          static_cast<E>(elt));
     }
 
 #ifdef USE_XSIMD

--- a/pythran/pythonic/include/types/numpy_texpr.hpp
+++ b/pythran/pythonic/include/types/numpy_texpr.hpp
@@ -86,12 +86,13 @@ namespace types
     {
       return arg.load(j, i);
     }
-    void store(dtype elt, long i, long j)
+    template <class Elt>
+    void store(Elt elt, long i, long j)
     {
       arg.store(elt, j, i);
     }
-    template <class Op>
-    void update(dtype elt, long i, long j) const
+    template <class Op, class Elt>
+    void update(Elt elt, long i, long j) const
     {
       arg.template update<Op>(elt, j, i);
     }

--- a/pythran/pythonic/include/types/numpy_vexpr.hpp
+++ b/pythran/pythonic/include/types/numpy_vexpr.hpp
@@ -76,13 +76,13 @@ namespace types
     {
       return data_.load(view_.fast(i), indices...);
     }
-    template <class... Indices>
-    void store(dtype elt, long i, Indices... indices) const
+    template <class Elt, class... Indices>
+    void store(Elt elt, long i, Indices... indices) const
     {
       data_.store(elt, view_.fast(i), indices...);
     }
-    template <class Op, class... Indices>
-    void update(dtype elt, long i, Indices... indices) const
+    template <class Op, class Elt, class... Indices>
+    void update(Elt elt, long i, Indices... indices) const
     {
       data_.template update<Op>(elt, view_.fast(i), indices...);
     }

--- a/pythran/tests/cases/train_equalizer.py
+++ b/pythran/tests/cases/train_equalizer.py
@@ -1,0 +1,48 @@
+import numpy as np
+
+#pythran export train_equaliser(complex128[][], int, int, int, float64, complex128[][][], int[], bool, complex128[][], str)
+#pythran export train_equaliser(complex64[][], int, int, int, float32, complex64[][][], int[], bool, complex64[][], str)
+def train_equaliser(E, TrSyms, Niter, os, mu, wx, modes, adaptive, symbols,  method):
+   if method == "cma":
+       errorfct = cma_error
+
+   else:
+       raise ValueError("Unknown method %s"%method)
+   nmodes = E.shape[0]
+   ntaps = wx.shape[-1]
+   assert symbols.shape[0] == nmodes, "symbols must be at least size of modes"
+   assert wx.shape[0] == nmodes, "wx needs to have at least as many dimensions as the maximum mode"
+   assert E.shape[1] > TrSyms*os+ntaps, "Field must be longer than the number of training symbols"
+   assert modes.max() < nmodes, "Maximum mode number must not be higher than number of modes"
+   err = np.zeros((nmodes, TrSyms*Niter), dtype=E.dtype)
+   #omp parallel for
+   for mode in modes:
+       for it in range(Niter):
+           for i in range(TrSyms):
+               X = E[:, i * os:i * os + ntaps]
+               Xest = apply_filter(X,  wx[mode])
+               err[mode, it*Niter+i] = errorfct(Xest, symbols[mode], i)
+               wx[mode] += mu * np.conj(err[mode, it*Niter+i]) * X
+               if adaptive and i > 0:
+                   mu = adapt_step(mu, err[mode, it*Niter+i], err[mode, it*Niter+i-1])
+   return err, wx, mu
+
+def cma_error(Xest, s1, i):
+   d = s1[0].real - abs(Xest)**2
+   return d*Xest
+
+
+def adapt_step(mu, err_p, err):
+    if err.real*err_p.real > 0 and err.imag*err_p.imag > 0:
+        return mu
+    else:
+        return mu/(1+mu*(err.real*err.real + err.imag*err.imag))
+
+def apply_filter(E, wx):
+    pols = E.shape[0]
+    Ntaps = wx.shape[1]
+    Xest = E.dtype.type(0)
+    for k in range(pols):
+        for i in range(Ntaps):
+            Xest += E[k, i]*np.conj(wx[k,i])
+    return Xest

